File: src/main/java/com/smarterd/collaboration/document/SharedDocumentEngine.java
package com.smarterd.collaboration.document;

import org.springframework.lang.NonNull;

/**
 * shared authoritative document engine 추상화.
 */
public interface SharedDocumentEngine {
    /**
     * 엔진 식별자를 반환한다.
     *
     * @return engine ID
     */
    @NonNull
    String engineId();

    /**
     * snapshot으로 엔진 상태를 hydrate한다.
     *
     * @param snapshot in-memory snapshot
     */
    void hydrate(@NonNull byte[] snapshot);

    /**
     * 현재 상태를 snapshot으로 export한다.
     *
     * @return in-memory snapshot
     */
    @NonNull
    byte[] exportSnapshot();

    /**
     * 원격 delta를 적용한다.
     *
     * @param delta remote delta
     */
    void applyRemoteDelta(@NonNull byte[] delta);
}

File: client/src/collaboration/core/contracts/shared-document-engine.ts
export interface SharedDocumentEngineUpdate {
  origin: unknown;
  changeSet?: unknown;
}

export interface SharedDocumentEngine {
  engineId: string;
  hydrate(snapshot: Uint8Array): void;
  exportSnapshot(): Uint8Array;
  applyRemoteDelta(delta: Uint8Array): void;
  subscribeUpdates(listener: (update: SharedDocumentEngineUpdate) => void): () => void;
}
